### ISSUE TITLE
Firefox 65 table scroll by keyboard

### ISSUE THREAD

**Issue by mappopo** [qooxdoo/qooxdoo#9623]
In Firefox 65 is no longer possible to change the row selection by keyboard.

The behaviour can be replicated in playgroud http://www.qooxdoo.org/current/demobrowser/#table~Table.html 
Select a row with a mouse click and try use keyboard arrows up and down to move selection. It takes no effect. With other browser (eg. Edge or Chrome) the focus and selection successful changes.

**Comment by zaucker** [comment]
Works for me on

   http://www.qooxdoo.org/devel/demobrowser/#table~Table.html

and

    http://www.qooxdoo.org/devel/playground/#Table-ria

with Ubuntu and FF 65

Cheers,
Fritz

On Tue, 5 Feb 2019, mappopo wrote:

> In Firefox 65 is no longer possible to change the row selection by keyboard.
>
> The behaviour can be replicated in playgroud http://www.qooxdoo.org/current/demobrowser/#table~Table.html
> Select a row with a mouse click and try use keyboard arrows up and down to move selection. It takes no effect. With other browser (eg. Edge or Chrome) the focus and selection successful changes.

-- 
Oetiker+Partner AG              tel: +41 62 775 9903 (direct)
Fritz Zaucker                        +41 62 775 9900 (switch board)
Aarweg 15                            +41 79 675 0630 (mobile)
CH-4600 Olten                   fax: +41 62 775 9905
Schweiz                         web: www.oetiker.ch

**Comment by cboulanger** [comment]
Works on MacOS 10.14.2, FF 65.

**Comment by mappopo** [comment]
I use windows

**Comment by hkollmann** [comment]
Yes it is. Most of us use git repos version

**Comment by cboulanger** [comment]
I see - I can confirm, http://www.qooxdoo.org/current/demobrowser/#table~Table.html is broken for Firefox 65 on the Mac. Looks like we need a backport of the fix?

**Comment by level420** [comment]
@mappopo did it work with firefox 64?

**Comment by lucapivato** [comment]
https://www.fxsitecompat.com/en-CA/docs/2018/non-printable-keys-no-longer-fire-keypress-event/

> An exception is the Enter key; pressing Enter with no modifiers, Shift+Enter or Ctrl+Enter will fire the keypress event as before, which is invalid in the current spec but compatible with other browsers.

This genius change may be coming to Chrome too...

**Comment by level420** [comment]
@lucapivato yes, currently investigating that and why it still works in master.

**Comment by level420** [comment]
The fix is https://github.com/qooxdoo/qooxdoo/pull/216 
@mappopo could you please try the fix in the PR above and report back? Thank you!

**Comment by level420** [comment]
created https://github.com/qooxdoo/qooxdoo/pull/9624 with cherry picked PRs #216 and #9404 from master

**Comment by mappopo** [comment]
> 
> 
> The fix is PR #216
> @mappopo could you please try the fix in the PR above and report back? Thank you!

it works, thank you

**Comment by adrelino** [comment]
In Firefox' 65 released on January 29th 2019, my legacy built on top of qooxdoo doesn't work properly anymore in many places since I haven't been maintaining it (especially with regard to events) properly.

Is it save to just replace `keypress` with `keydown` and `_onKeyPress` with `_onKeyDown`  (used e.g. in our own Table subclass) everywhere in my own code which extends qooxdoo?

Even though keypress still works in chrome, it seems to be officially deprecated now (https://developer.mozilla.org/en-US/docs/Web/Events/keypress, https://w3c.github.io/uievents/#event-type-keypress) and even Mozilla suggests to replace it with `beforeinput` or `keydown`. Or are there any downsides expected from this?

**Comment by adrelino** [comment]
While searching for "keypress", I just stumbled upon another related bug inside qooxdoo code.

For the same reason, even in the current devel version, one cannot close an opened SelectBox anymore by pressing Escape in Firefox 65, while in Chrome 71 it works:

http://www.qooxdoo.org/devel/demobrowser/#widget~SelectBox.html

The reason is that AbstractSelectBox is still listening for `keypress` instead of `keydown`.

```
// Register listeners
this.addListener("keypress", this._onKeyPress);
this.addListener("blur", this._onBlur, this);
```
[source](https://github.com/qooxdoo/qooxdoo/blob/7cbcf73fe38e650daec33b639a028d9904b57a09/framework/source/class/qx/ui/form/AbstractSelectBox.js#L59)

**Comment by adrelino** [comment]
Yet another related issue:
Controlling a RadioButton with the arrow keys works in chrome 72, while in Firefox 65:
- doesn't work in [current](http://www.qooxdoo.org/current/demobrowser/#widget~RadioButton.html)
- whereas in [devel](http://www.qooxdoo.org/devel/demobrowser/#widget~RadioButton.html), one arrow key press triggers moving to the next radio item 2 times (so you get 2 alerts) and end up 2 options further...

**Comment by level420** [comment]
@adrelino Thank you for reporting this. From Firefox 65 on there's no keypress event sent anymore for non printable characters. See https://www.fxsitecompat.com/en-CA/docs/2018/non-printable-keys-no-longer-fire-keypress-event/

The solution is indeed to switch over to use the keydown event in this situation.

It would be great if you can offer a PR for master which I'll then backport, if applicable, to branch 5_0_x.

I think generating artificial keypress events for browsers not sending native keypress anymore isn't an option, as it seems that browsers will in the end follow step by step the specs and in the end deprecate keypress (https://w3c.github.io/uievents/#keypress)

**Comment by oetiker** [comment]
I wonder if we could put something in place to warn authors about this problem ...maybe have the compailer complain if it sees keypress eventhandlers being assigned ? @johnspackman

**Comment by level420** [comment]
@oetiker I think we should switch over to #9626 with our discussion

**Comment by level420** [comment]
@adrelino the problem with not closing the select box popup via escape in FF 65 should be solved with PR https://github.com/qooxdoo/qooxdoo/pull/9629 

The problem of the double event when navigating in radio buttons in the demo http://cpr2/qooxdoo-level420/application/demobrowser/source/#widget~RadioButton.html is still there, but it may be another problem where the event somehow interacts with the alert popup displayed on value change. There is no problem in navigating with the arrow keys in the demo http://cpr2/qooxdoo-level420/application/demobrowser/source/#widget~RadioButtonGroup.html where no alert box is displayed on value change.
